source/s3_default_meta_request.c

```diff
@@ -208,20 +219,82 @@ static bool s_s3_meta_request_default_update(
 }
 
 /* Given a request, prepare it for sending based on its description. */
-static int s_s3_meta_request_default_prepare_request(
-    struct aws_s3_meta_request *meta_request,
-    struct aws_s3_request *request) {
+static struct aws_future *s_s3_default_prepare_request_async(struct aws_s3_request *request) {
+    AWS_PRECONDITION(request);
+
+    struct aws_s3_meta_request *meta_request = request->meta_request;
     AWS_PRECONDITION(meta_request);
 
     struct aws_s3_meta_request_default *meta_request_default = meta_request->impl;
     AWS_PRECONDITION(meta_request_default);
 
+    struct aws_future *preparation_future = aws_future_new(request->allocator, AWS_FUTURE_VALUELESS);
+
+    /* Store context for async operation */
+    struct aws_s3_default_prepare_request_async_ctx *request_prep =
+        aws_mem_calloc(request->allocator, 1, sizeof(struct aws_s3_default_prepare_request_async_ctx));
+    request_prep->allocator = request->allocator;
+    request_prep->request = request;
+    request_prep->future = aws_future_acquire(preparation_future);
+
     if (meta_request_default->content_length > 0 && request->num_times_prepared == 0) {
         aws_byte_buf_init(&request->request_body, meta_request->allocator, meta_request_default->content_length);
 
-        if (aws_s3_meta_request_read_body(meta_request, &request->request_body)) {
-            return AWS_OP_ERR;
-        }
+        /* Kick off the async read */
+        struct aws_future *read_future = aws_s3_meta_request_read_body(meta_request, &request->request_body);
+        aws_future_register_callback(read_future, s_s3_default_prepare_request_on_read_done, request_prep);
```

Comment:
Debatable: I'd prefer that we release the 'future' here and allow 'aws_future_register_callback' to acquire the reference before the callback, and then release it after the callback. This would eliminate the need to release it within the callback.

Reply:
We talked. The future callback no longer passes a pointer to itself. This makes the ownership model more obvious.  Now the consumer MUST store a pointer to the future or they won't have access to it from the callback. Now the consumer generally releases the reference to that future when they clean up everything else stored in their `async_ctx` struct.